### PR TITLE
arCalibration Merge

### PR DESCRIPTION
**Pull request by ComAladar**
Рефакторинг; удаление ненужного.

### CODE REVIEW

File: app/src/main/java/com/example/synclient/ui/ar/ARCameraActivity.kt [hunk 1]
```diff
@@ -58,59 +63,15 @@ class ARCameraActivity : AppCompatActivity() {
     }
 
     private fun updateConfig() {
-        while(arFragment.arSceneView.session == null){
+        while (managerAR.arFragment.arSceneView.session == null) {
```
Comment (EssenceShift) [inline app/src/main/java/com/example/synclient/ui/ar/ARCameraActivity.kt]: Семен сказал, что так нагрузка чуть выше и лучше оставить скобки пустыми, не принципиально, но лучше убрать.

Comment (pradlol) [reply app/src/main/java/com/example/synclient/ui/ar/ARCameraActivity.kt]: вообще, если докопаться, то выхлоп идентичен. синтетический, но всё-таки пруф: https://godbolt.org/z/eE7fsvY94

File: app/src/main/java/com/example/synclient/arLogic/ManagerAR.kt [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+import com.google.ar.sceneform.ux.ArFragment
+
+/**
+ * Класс ManagerAR
+ *
+ * Класс для работы и хранения всех виджетов и обьектов для работы с ARCore.
+ * @param context Конекст основной активити.
+ * @constructor Создает экземпляр класса с передаваемым в него контекстом.
+ */
+class ManagerAR constructor(context: Context) {
```
Comment (pradlol) [inline app/src/main/java/com/example/synclient/arLogic/ManagerAR.kt]: 1. Имя класса не говорит о том, что оно делает. "Manager" в целом в имени класса считается плохим тоном (есть минираздел в книги Р. Мартина Чистый код, есть [такой](https://softwareengineering.stackexchange.com/questions/129537/can-manager-classes-be-a-sign-of-bad-architecture) ответ). 

2. У таких сущностей есть все шансы вырости до антипаттерна Божественный объект ([ляпс](https://ru.wikipedia.org/wiki/%D0%91%D0%BE%D0%B6%D0%B5%D1%81%D1%82%D0%B2%D0%B5%D0%BD%D0%BD%D1%8B%D0%B9_%D0%BE%D0%B1%D1%8A%D0%B5%D0%BA%D1%82) [ляпс](https://refactoring.guru/ru/smells/large-class))

3. ARCameraActivity, в половине случаев обращается не к ManagerAr, а к его содержимому, что является антипаттерном [завистливые функции](https://refactoring.guru/ru/smells/feature-envy)
`if (managerAR.anchor.trackingState == TrackingState.PAUSED) { // works with anchor`
`managerAR.anchor.detach() // works with anchor`
`var config = managerAR.arFragment.arSceneView.session?.config // works with arFragment and its contents`
`managerAR.arFragment.arSceneView.session?.configure(config) // works with arFragment and its contents`
`while (managerAR.arFragment.arSceneView.session == null) { // works with arFragment and its contents`

File: app/src/main/java/com/example/synclient/arLogic/PortAR.kt [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+import com.google.ar.sceneform.rendering.ViewRenderable
+import com.google.ar.sceneform.ux.ArFragment
+import com.google.ar.sceneform.ux.TransformableNode
+
+/**
+ * Класс PortAR
+ *
+ * Класс для создания и отображения обьекта порта устройства.
+ */
+class PortAR {
```
Comment (pradlol) [inline app/src/main/java/com/example/synclient/arLogic/PortAR.kt]: Имя класса не отражает его деятельность, и больше похож на тот же билдер

File: app/src/main/java/com/example/synclient/arLogic/WidgetAR.kt [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+import com.google.ar.sceneform.rendering.ViewRenderable
+import com.google.ar.sceneform.ux.ArFragment
+import com.google.ar.sceneform.ux.TransformableNode
+
+/**
+ * Класс WidgetAR
+ *
+ * Класс для создания и отображения основного виджета с информацией об устройстве.
+ */
+class WidgetAR {
```
Comment (pradlol) [inline app/src/main/java/com/example/synclient/arLogic/WidgetAR.kt]: именование класса очень абстрактное, что за виджет (без просмотра кода содержимого) непонятно. Судя по комментарию, класс больше DeviceInfoView; судя по содержимому, класс больше DeviceInfoViewBuilder.

File: app/src/main/java/com/example/synclient/arLogic/ManagerAR.kt [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+                    port.createPort(
+                        arFragment,
+                        anchor,
+                        1,
+                        myContext,
+                        portsCoords[1][0],
+                        portsCoords[1][1],
+                        portsCoords[1][2]
+                    )
+                    portList.toMutableList().add(port)
```
Comment (pradlol) [inline app/src/main/java/com/example/synclient/arLogic/ManagerAR.kt]: portList и так объявлен как mutable

File: app/src/main/java/com/example/synclient/arLogic/ManagerAR.kt [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+                    port.createPort(
+                        arFragment,
+                        anchor,
+                        0,
+                        myContext,
+                        portsCoords[0][0],
+                        portsCoords[0][1],
+                        portsCoords[0][2]
+                    )
+                    portList.toMutableList().add(port)
```
Comment (pradlol) [inline app/src/main/java/com/example/synclient/arLogic/ManagerAR.kt]: для 16 портов будет аналогичное создание?

Comment (ComAladar) [reply app/src/main/java/com/example/synclient/arLogic/ManagerAR.kt]: Вроде исправил ошибку. Уже закоммитил.

File: app/src/main/java/com/example/synclient/arLogic/WidgetAR.kt (hunk 4 above)
Comment (ComAladar) [reply app/src/main/java/com/example/synclient/arLogic/WidgetAR.kt]: Переименовал классы, но все равно не совсем понятна в данном примере разница между DeviceInfoView и Builder. Как я понимаю, View должен бы был просто работать с данными и их отображением а Builder именно с отображением самого виджета в AR. Но мы с данными никак не работаем, поэтому назвал пока что Builder, но если что поменяю.

File: app/src/main/java/com/example/synclient/arLogic/PortAR.kt (hunk 3 above)
Comment (ComAladar) [reply app/src/main/java/com/example/synclient/arLogic/PortAR.kt]: Поменял наименование класса.

File: app/src/main/java/com/example/synclient/arLogic/ManagerAR.kt (hunk 2 above)
Comment (ComAladar) [reply app/src/main/java/com/example/synclient/arLogic/ManagerAR.kt]: Понял в чем основная проблема, но не знаю как их правильно решить. Если по пунктам:
1) Как иначе назвать класс, который в себе содержит экземпляры остальных? Разбивать его на несколько бессмысленно(хотя это возможно, но тогда нам в каждый класс(port итд) придется передавать anchor, arFragment итд, Возможно стоит переделать ManagerAR чтобы он просто создавал anchor, а ARCameraActivity как и раньше будет хранить все экземпляры, но я менеджера и делал чтобы этого избежать,
2) Согласен, но как этого избежать непонятно. Класс создан изначально для упрощения всей работы с AR чтобы не надо было вызывать методы из разных классов которые по сути делают одно и тоже. Если один класс отвечает за всю работу с AR это будет божественный обьект или нет? В данный момент не уверен, так что пока оставил как есть.
3) Исправил другую часть кода, вроде работает но там ARCore обновился и у меня эмулятор падает, поэтому завтра проверю и посмотрим. Если как я сделал работает, то перепишу оставшуюся часть. Весь этот кусок из ARCameraActivity затолкал в ManagerAR и наверно это еще больше приводит к пункту 2) но тут уже не знаю)

File: app/src/main/java/com/example/synclient/ui/ar/ARCameraActivity.kt (hunk 1 above)
Comment (EssenceShift) [reply app/src/main/java/com/example/synclient/ui/ar/ARCameraActivity.kt]: Да, ты прав.
Без оптимизации: https://godbolt.org/z/Pf7fK4fad - появляется лишняя метка
С оптимизацией: https://godbolt.org/z/sMP65b8a4
Но компилятор достаточно умный, так что это и правда бесполезно.

File: app/src/main/java/com/example/synclient/arLogic/ManagerAR.kt (hunk 2 above)
Comment (pradlol) [reply app/src/main/java/com/example/synclient/arLogic/ManagerAR.kt]: 1. Если не ошибаюсь, ARCameraActivity так или иначе содержит в себе все классы, с которыми работает ManagerAR. Если ты оставишь в менеджере только метод создания якоря, то это уже похоже на [ленивый класс](https://refactoring.guru/ru/smells/lazy-class). Чтобы сказать наверняка, нужно сесть вместе и попроектировать. Сейчас времени у вас не так уж и много осталось, так что можем отложить вопрос, отрефакторить всегда успеется.


2. Под всей работой с AR можно ведь рассматривать: 
* конфигурация фреймворка arcore (возня с конфигами, ауг имейдж, ...);
* обрабокта найденной пикчи с созданием контролов;
* обработка действий пользователя;
* некая бизнес логика, переход между стейтами.

Если всё это выполняет ManagerAR, то это уже несомненно божество) Опять же, сейчас тогда можем не заострять на этом внимание.

**Review by pradlol — APPROVED** [review]
